src/scf/newton.jl

```diff
@@ -0,0 +1,420 @@
+
+Compute the residual associated to a set of planewave φ, that is to say
+H(φ)*φ - λ*φ where λ is the set of rayleigh coefficients associated to the φ.
+"""
+function compute_residual(basis::PlaneWaveBasis, φ, occ)
+
+    # necessary quantities
+    Nk = length(basis.kpoints)
+    ρ = compute_density(basis, φ, occ)
+    energies, H = energy_hamiltonian(basis, φ, occ; ρ=ρ)
+
+    # compute residual
+    res = similar(φ)
+    for ik = 1:Nk
+        φk = φ[ik]
+        N = size(φk, 2)
+        Hk = H.blocks[ik]
+        # eigenvalues as rayleigh coefficients
+        egvalk = φk'*(Hk*φk)
+        # compute residual at given kpoint as H(φ)φ - λφ
+        res[ik] = Hk*φk - φk*egvalk
+    end
+    res
+end
+
+# To project onto the space tangent to φ, we project δφ onto the orthogonal of φ
+function proj_tangent(δφ, φ; tol_test=1e-12)
+
+    Nk1 = size(δφ,1)
+    Nk2 = size(φ,1)
+    @assert Nk1 == Nk2
+    Nk = Nk1
+
+    Πδφ = similar(δφ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = deepcopy(δφk)
+
+        N1 = size(δφk,2)
+        N2 = size(φk,2)
+        @assert N1 == N2
+        N = N1
+
+        for i = 1:N, j = 1:N
+            Πδφk[:,i] -= (φk[:,j]'δφk[:,i]) * φk[:,j]
+        end
+        Πδφ[ik] = Πδφk
+    end
+
+    # test orthogonalisation
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = Πδφ[ik]
+        N = size(φk,2)
+        for i = 1:N, j = 1:N
+            @assert abs(Πδφk[:,i]'φk[:,j]) < tol_test [println(abs(Πδφk[:,i]'φk[:,j]))]
+        end
+    end
+
+    Πδφ
+end
+
+# KrylovKit custom orthogonaliser to be used in KrylovKit eigsolve, svdsolve,
+# linsolve, ...
+# This has to be passed in KrylovKit solvers when studying (Ω+K) as a super
+# operator from the tangent space to the tangent space to be sure that
+# iterations are constrained to stay on the tangent space.
+# /!\ Be careful that here, φ needs to be packed into on big array
+struct OrthogonalizeAndProject{F, O <: Orthogonalizer, φ} <: Orthogonalizer
+    projector::F
+    orth::O
+    φ::φ
+end
+OrthogonalizeAndProject(projector, φ) = OrthogonalizeAndProject(projector,
+                                                                KrylovDefaults.orth,
+                                                                φ)
+function KrylovKit.orthogonalize!(v::T, b::OrthonormalBasis{T}, x::AbstractVector,
+                                        alg::OrthogonalizeAndProject) where {T}
+    v, x = orthogonalize!(v, b, x, alg.orth)
+    v = reshape(v, size(alg.φ))
+    v = alg.projector(v, alg.φ)::T
+    v, x
+end
+function KrylovKit.orthogonalize!(v::T, x::AbstractVector,
+                                        alg::OrthogonalizeAndProject) where {T}
+    v, x = orthogonalize!(v, x, alg.orth)
+    v = reshape(v, size(alg.φ))
+    v = alg.projector(v, alg.φ)::T
+    v, x
+end
+function KrylovKit.gklrecurrence(operator, U::OrthonormalBasis, V::OrthonormalBasis, β,
```

Comment:
Is this stuff used? None of this seems to be covered by tests.

Reply:
I think I needed to implement it for eigen and svd solvers in KrylovKit.

Reply:
But not strictly necessary for newton indeed, maybe we can skip it for the moment or move it in another file where we compute the operator norms.

Reply:
I can add a test, but I'm not sure what I can test. I can compute svd of Ω+K, but what should I check then ? Just that the code runs ? Cause I'm not sure if I can add a test like I can verify the value of a given singular value as we need small Ecut to make compuations fast...

Reply:
test for non-zero smallest singular value